sdk/python/feast/job_service.py

```diff
@@ -145,6 +167,54 @@ def GetJob(self, request, context):
         return GetJobResponse(job=self._job_to_proto(job))
 
 
+def run_control_loop():
+    """Runs control loop that continuously ensures that correct jobs are being run."""
+    logging.info(
+        "Feast Job Service is starting a control loop in a background thread..."
+    )
+    client = feast.Client()
+    while True:
+        # Map of job hash -> (project, table_name)
+        table_metadata: Dict[str, Tuple[str, str]] = dict()
+        # Job hashes that should exist
+        final_job_hashes: Set[str] = set()
+
+        for project in client.list_projects():
```

Comment:
I think this `run_control_loop` function can be written a little bit cleaner by pulling out some parts of these deeply nested loops into their own private methods

Reply:
Just as another 2c, at least until this becomes too unwieldy, I would rather err on the side of having more logic inline rather than extracting it into helper functions here ([see this](http://number-none.com/blow/blog/programming/2014/09/26/carmack-on-inlined-code.html) )